### DISCUSSION TITLE
refreshing countdown timer with timeline sections

### DISCUSSION THREAD

**Discussion by abhilasha-kumar** [jspsych/jsPsych#2673] [Q&A] [answered]
Hi, I am programming a simple experiment where I would like participants to list all of the "X" they can think of in 3 mins, where X changes in each block (e.g., animals, foods, etc.). I want to display a countdown on the screen for each **block** and after 3 minutes have passed, I would like the loop to end and move to the next domain (e.g., foods). see screenshot below:

<img width="1350" alt="Screen Shot 2022-06-21 at 4 13 29 PM" src="https://user-images.githubusercontent.com/26175039/174889186-252593d1-9c27-416f-9f67-95021240632f.png">

I was able to display a timer and make it work with the jsPsych plugins using some of the previous discussion threads. I have two issues:

1. The timer starts when the instruction screen is on, but I want it to start only when the loop starts and not be displayed otherwise. I think this is an issue with the `on_load` argument as I imagine it is called as soon as the script is read, but I don't know how to make it so that the timer only runs when the trial starts.
2. Is there a way to end the loop using the timer info directly? currently, I am using a separate function for displaying the countdown and a separate time calculation to end the loop, which seems inefficient and imprecise, and also doesn't automatically end when the timer stops (it waits for the last input to end the loop).

The experiment is available [here](https://github.com/abhilasha-kumar/words-demo) for a quick demo.
Hope someone can help, thanks!

@jodeleeuw any thoughts on this?

**Comment by jodeleeuw** [comment, marked as answer]
Hi @abhilasha-kumar,

The timer is starting before the loop because you are setting `on_load: countDown()`, which invokes the `countDown` function immediately on page loading due to the `()`. You can instead pass the function without calling it, and it should then be called when the plugin loads: `on_load: countDown`.

As for ending the loop using the timer info directly, it should be possible but it will be a little messy. You can call `jsPsych.finishTrial()` to end the current trial. However, this wasn't designed for use outside of plugins, so calling it won't prevent the plugin from *also* calling it, especially if the plugin uses some kind of timer mechanism to trigger the call. You therefore might need to call things like `jsPsych.pluginAPI.clearAllTimeouts()` to prevent that from happening.

**Reply by abhilasha-kumar** [reply]
Hi @jodeleeuw! Thanks for your input. I tried to remove the `()` from `on_load`, but it now seems like the timer restarts on each individual trial, as opposed to continuing until the full block is over. Basically, I want participants to type as many responses as they can until the timer runs out. But I think what's happening is that the countDown function is being called from the beginning as soon as a response is made. I thought that embedding the individual trial inside a loop would solve this but no luck, see below:

```javascript
var animal = {
    type: jsPsychSurveyText,
    questions: [
    {prompt: '<p style="font-size:40px;">List all the ANIMALS you can in 3 minutes.</p>'}
  ]
  }

var animal_loop = {
    on_load: countDown,
    timeline: [animal],
    loop_function: function(data){
      var getTime = new Date().getTime();
      var timePassedCounter = getTime - startTime;

        if(timePassedCounter < 15000){
          console.log("timePassed: " + timePassedCounter);
            return true;
        } else {
          console.log("timePassed: " + timePassedCounter);
            startTime = new Date().getTime();
            return false;
        }
    }
}
```

Let me know if you have thoughts on this, thanks!

**Reply by jodeleeuw** [reply]
I think when you define `on_load` at the level of a timeline (as in your `animal_loop`) then jsPsych recursively applies that function to every trial on that timeline. If you want the function to run only when the timeline starts then you can use `on_timeline_start` documented [here](https://www.jspsych.org/7.2/overview/timeline/#timeline-start-and-finish-functions). However, I think that `on_timeline_start` may run every time the loop restarts (I actually don't remember this, so please check 😁). If that's the case, then I think the simplest fix would be to add a `jsPsychCallFunction` trial right before the `animal_loop` starts to invoke the `countDown` function.

**Reply by jodeleeuw** [reply]
BTW, I think a custom plugin that implemented a timed recall task like this would be a great addition to the library if you were interested in moving the work inside a plugin 👍

**Reply by abhilasha-kumar** [reply]
Thanks @jodeleeuw! using the `jsPsychCallFunction` did the job! And yes I agree, having this as a plugin would be great - adding it to my to-do list 💯

**Reply by AnneScholz** [reply]
Hi, I just stumbled across this thread, because I would also like to use a timer over several trials. I'm an absolute newbie to jsPsych and therefore don't understand how you can use the jsPsychCallFunction to solve this problem. I have been trying a couple of options, but neither of them work. The timer keeps resetting after completing a trial. Would you be willing to share the code which worked out for you in the end? Any help would be highly appreciated :)
